salvo/src/lib/source_manager.py

```diff
@@ -135,4 +271,114 @@ def get_source_repository(self, \
+
+    if not source:
+      raise SourceManagerError(
+          f"Unable to find a source with the requested ID: {source_name}")
+
+    return source
+
+  def _create_source_tree(self, \
+      source_id: proto_source.SourceRepository.SourceIdentity) \
+      -> source_tree.SourceTree:
+    """Creates a source tree object from a SourceRepository.
+
+    Args:
+      source_id: The identity of the source object we seek (eg.
+        SRCID_NIGHTHAWK or SRCID_ENVOY)
+    Returns:
+      a source tree object managing the identified source repository
+    """
+
+    repo = self.get_source_repository(source_id)
+    return source_tree.SourceTree(repo)
+
+  def get_source_tree(self, \
+      source_id: proto_source.SourceRepository.SourceIdentity) \
+      -> source_tree.SourceTree:
+    """Returns the source tree object identified by source_id.
+
+    Args:
+      source_id: The identity of the source tree we seek (eg.
+        SRCID_NIGHTHAWK or SRCID_ENVOY)
+
+    Returns:
+      a source tree object managing the identified source repository
+
+    Raises:
+      SourceManagerError if no source tree is found
+    """
+
+    if source_id not in self._source_tree:
+      source_name = proto_source.SourceRepository.SourceIdentity.Name(source_id)
+      raise SourceManagerError(
+          f"No Source tree defined for: {source_name}")
+
+    return self._source_tree[source_id]
+
+  def get_build_options(self, \
+      source_id: proto_source.SourceRepository.SourceIdentity) \
+      -> proto_source.BazelOption:
+    """Determine whether build options are specified in the control object
+    and return them
+
+    Args:
+      source_id: The identity of the source object we seek (eg.
+        SRCID_NIGHTHAWK or SRCID_ENVOY)
+
+    Return:
+      the Bazel Options defined in the source identified by the
+        specified source_id
+
+    Raises:
+      SourceManagerError: If no options are defined in the source object
+    """
+
+    source = self.get_source_repository(source_id)
+    bazel_options = source.bazel_options
+
+    if not bazel_options:
+      source_name = proto_source.SourceRepository.SourceIdentity.Name(source_id)
+      raise SourceManagerError(
+          f"No Bazel Options are defined in source: {source_name}")
+
+    return bazel_options
+
+  def have_build_options(self, \
+      source_id: proto_source.SourceRepository.SourceIdentity) \
+      -> bool:
+    """Determine whether build options are specified in the control object
+       and return a boolean.  This is used to determine whether we build
+       images or use the already available images
+
+    Args:
+      source_id: The identity of the source object we seek (eg.
+        SRCID_NIGHTHAWK or SRCID_ENVOY)
+
+    Return:
+      a boolean indicating the presense of user specified bazel options
+    """
+    try:
+      build_options = self.get_build_options(source_id)
+      options_present = len(build_options) >= 1
```

Comment:
This is the other uncovered line which too seems to represent non-trivial behavior. It looks like our tests only use one path through this logic - the one that raises an exception. Can we cover the non-exception path too?

Reply:
Updated and verified that we're hitting these lines in the html report :)

Reply:
An exceptional test coverage ;) Thank you.